sda-commons-shared-asyncapi/src/main/java/org/sdase/commons/shared/asyncapi/internal/JsonSchemaEmbedder.java

```diff
@@ -0,0 +1,138 @@
+package org.sdase.commons.shared.asyncapi.internal;
+
+import com.fasterxml.jackson.core.JsonPointer;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.JsonNodeFactory;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+/** Embeds external references in a JSON schema using a schema resolver. */
+public class JsonSchemaEmbedder {
+
+  private static final String REF = "$ref";
+  private static final List<String> IGNORE_ON_EMBED = Arrays.asList("$schema", "definitions");
+
+  private final JsonPointer embedPointer;
+  private final JsonSchemaResolver jsonSchemaResolver;
+
+  /**
+   * Create a new JSON schema embedder.
+   *
+   * @param embedPath The JSON path where the JSON schemas should be embedded.
+   * @param jsonSchemaResolver Resolver for external schema files.
+   */
+  public JsonSchemaEmbedder(String embedPath, JsonSchemaResolver jsonSchemaResolver) {
+    this.embedPointer = JsonPointer.compile(embedPath);
+    this.jsonSchemaResolver = jsonSchemaResolver;
+  }
+
+  /**
+   * Resolves external references in the input JSON object.
+   *
+   * @param input The input JSON object containing external references.
+   * @return A cloned JSON object, that contains all external references that could be resolved,
+   *     embedded a the embed path.
+   */
+  public JsonNode resolve(JsonNode input) {
+    Map<String, JsonNode> resolvedReferences = new HashMap<>();
+    JsonNode output = input.deepCopy();
+
+    resolveInPlace(output, output, output, resolvedReferences, false);
+
+    return output;
+  }
+
+  private void resolveInPlace(
+      JsonNode target,
+      JsonNode root,
+      JsonNode current,
+      Map<String, JsonNode> resolvedReferences,
+      boolean shouldEmbedInternal) {
+    if (current.isArray()) {
+      current.forEach(
+          jsonNode ->
+              resolveInPlace(target, root, jsonNode, resolvedReferences, shouldEmbedInternal));
+    } else if (current.isObject()) {
+      List<Entry<String, JsonNode>> fields = new ArrayList<>();
+      current.fields().forEachRemaining(fields::add);
+
+      fields.forEach(
+          e -> {
+            if (REF.equals(e.getKey()) && e.getValue().isTextual()) {
```

Comment:
ignore case?

Reply:
I can't find a statement that keywords in json schema are case insensitive nor that they are case sensitive. However the underlining JSON standard is case sensitive (you can have both the properties `$ref` and `$REF` in the same document), so I think the keywords should be case sensitive.